### PR TITLE
clSetEventCallback & trace improvements

### PR DESCRIPTION
**Pull request by clementleger**
This commit fixes the way clSetEventCallback should work and introduce a tracing mecanism.
This trace system can be controlled by an ENV var (POCL_TRACE_EVENT) which will log all events.
The logging itself cabn be done by multiple backend (simple text logger at the moment) but will be extending to add a lttng tracer for instance.

### CODE REVIEW

**Comment by franz** [comment]
I guess you didn't notice (or perhaps you have some other reasons), but we already have a (slightly more advanced) "tracing" mechanism in pocl, in lib/CL/pocl_debug.{c,h}. It's used quite extensively across entire pocl codebase... it's enabled via ```POCL_DEBUG=1``` env var.

The current problem with it is, it can be really chatty, so we talked about improving it. I would like to introduce some sort of categories, like:
 ```
POCL_DEBUG={all,memory,timing,api,events,....}
```

So it's on the TODO list, just not done yet... when it's done the events category would do what your patch does now. So let me know what you think.

**Comment by clementleger** [comment]
The latest commits add support for POCL_TRACE_EVENT_FILTER env var in order to filter specific event statuses (complete,running,etc). It also add support for LTTNG traces.
As discussed on IRC, the goal is not the same as POCL_DEBUG (which will probably ben deactivated in production releases). Moreover, There is more informations on commands in order to easily plug a trace viewer or something similar and output clear diagrams.

File: lib/CL/pocl_tracing.c [hunk 1]
```diff
@@ -0,0 +1,299 @@
+#include <stdio.h>
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: This file lacks the license comment.

File: lib/CL/pocl_tracing.c [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+#ifdef LTTNG_UST_AVAILABLE
+  &lttng_tracer,
+#endif
+};
+
+#define POCL_TRACER_COUNT (sizeof(pocl_event_tracers) / sizeof((pocl_event_tracers)[0]))
+
+static const struct pocl_event_tracer *event_tracer = NULL;
+
+int pocl_event_updated(cl_event event, int status)
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Please check: http://www.gnu.org/prep/standards/html_node/Formatting.html#Formatting -- should have the return value in its own line etc.

File: lib/CL/pocl_tracing.c [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+{
+  event_callback_item* cb_ptr;
+
+  /* event callback handling just call functions in the same order
+   * they were added if the status match the specified one */
+  for (cb_ptr = event->callback_list; cb_ptr; cb_ptr = cb_ptr->next)
+    {
+      if (cb_ptr->trigger_status == status)
+        cb_ptr->callback_function (event, cb_ptr->trigger_status,
+						  cb_ptr->user_data);
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Let's use space indentation to avoid this type of "printing" issues.

File: lib/CL/pocl_tracing.c [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+  if (event_tracer && ((1 << status) & event_trace_filter))
+    event_tracer->event_updated(event, status);
+}
+
+void pocl_parse_event_filter()
+{
+  const char *trace_filter;
+  char *tmp_str, *save_ptr, *token;
+
+  trace_filter = pocl_get_string_option("POCL_TRACE_EVENT_FILTER", NULL);
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Check C style. Function call parenthesis like: func (foo).

File: lib/CL/pocl_tracing.c [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+    goto EVENT_INIT_OUT;
+
+  /* Check if a tracer has a name matching the supplied one */
+  for (i = 0; i < POCL_TRACER_COUNT; i++)
+    {
+	    if (strcmp(trace_env, pocl_event_tracers[i]->name) == 0)
+	      {
+					event_tracer = pocl_event_tracers[i];
+					break;
+				}
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Indent issue.

File: lib/CL/pocl_tracing.c [hunk 6]
```diff
@@ -0,0 +1,299 @@
...
+ */
+static FILE *text_tracer_file = NULL;
+static pocl_lock_t text_tracer_lock = POCL_LOCK_INITIALIZER;
+
+static void text_tracer_init()
+{
+  const char *text_tracer_output;
+
+  text_tracer_output = pocl_get_string_option("POCL_TRACE_EVENT_OPT", 
+											"pocl_trace_events.log");
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Indent.

File: lib/CL/pocl_tracing.c [hunk 7]
```diff
@@ -0,0 +1,299 @@
...
+  cl_ulong ts = cq->device->ops->get_timer_value(cq->device->data);
+  _cl_command_node *node = event->command;
+  char tmp_buffer[512];
+  char *cur_buf = tmp_buffer;
+  int text_size = 0;
+  
+  /* Some less integrated commands (clEnqueueReadBufferRect) do not use
+   * standard mecanism, so check node to be non null */
+  if (node == NULL)
+	  return;
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Indent.

File: lib/CL/pocl_tracing.c [hunk 8]
```diff
@@ -0,0 +1,299 @@
...
+
+  text_size = sprintf(cur_buf, "%lld %s %s ", ts,
+												command_to_str(event->command_type),
+												status_to_str[event->status]);
+	cur_buf += text_size;
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Indent.

File: lib/CL/pocl_tracing.c [hunk 9]
```diff
@@ -0,0 +1,299 @@
...
+  char *cur_buf = tmp_buffer;
+  int text_size = 0;
+  
+  /* Some less integrated commands (clEnqueueReadBufferRect) do not use
+   * standard mecanism, so check node to be non null */
+  if (node == NULL)
+	  return;
+
+  text_size = sprintf(cur_buf, "%lld %s %s ", ts,
+												command_to_str(event->command_type),
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Indent.

File: lib/CL/pocl_tracing.c [hunk 10]
```diff
@@ -0,0 +1,299 @@
...
+												status_to_str[event->status]);
+	cur_buf += text_size;
+	/* Print more informations for some commonly used commands */
+  switch (event->command_type)
+    {
+    case CL_COMMAND_READ_BUFFER:
+      text_size += sprintf(cur_buf, "size=%d, host_ptr=%p\n",
+														node->command.read.cb,
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.c]: Indent.

File: lib/CL/pocl_tracing.h [hunk 11]
```diff
@@ -0,0 +1,65 @@
+/* pocl_event.h: interface for event update and tracing system
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.h]: Check copy paste error pocl_event.h?

File: lib/CL/pocl_tracing.h [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+#ifndef POCL_EVENT_H
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.h]: Should be POCL_TRACING_H

File: lib/CL/pocl_tracing.h [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+#endif
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+int pocl_event_updated(cl_event event, int new_status);
+
+/**
+ * Initialized event tracing system
```
Comment (pjaaskel) [inline lib/CL/pocl_tracing.h]: -d

**Comment by pjaaskel** [comment]
Ungh. I think I accidentally wiped out my per-line comments? Summary: looking good except please check http://www.gnu.org/prep/standards/html_node/Formatting.html#Formatting in C files. Also, add a missing license text plus a copy paste error in pocl_events.h (in the license text and the #ifdef protect macro).

**Comment by clementleger** [comment]
Thanks Pekka, I fixed the coding style using `indent` with recommended options, it should be ok.
I also added the copyright to pocl_lttng.* files.

**Comment by pjaaskel** [comment]
LGTM now. @franz any remaining objections of me testing&merging this in?

File: lib/CL/pocl_tracing.c [hunk 14]
```diff
@@ -0,0 +1,350 @@
...
+    return;
+
+  event_trace_filter = 0;
+  while (1)
+    {
+      token = strtok_r (tmp_str, ",", &save_ptr);
+      if (token == NULL)
+	goto PARSE_OUT;
+      if (strcmp (token, "queued") == 0)
+	event_trace_filter |= (1 << CL_QUEUED);
```
Comment (clementleger) [inline lib/CL/pocl_tracing.c]: Hum it seems like indent added tabs here.

**Comment by franz** [comment]
Small idea - perhaps I would move status_to_str and command_to_str into pocl_util.{c,h}, they look like they could be useful in other code too.

Anyway, looks good to me.

**Comment by clementleger** [comment]
I agree, it could be used for pocl_debug in clFinish for instance. I will do that tomorrow.

**Pull request merged**